File: 542.01-matrix.cpp
#include <bits/stdc++.h>
using namespace std;
/*
 * @lc app=leetcode id=542 lang=cpp
 *
 * [542] 01 Matrix
 */

// @lc code=start
class Solution
{
public:
    vector<vector<int>> updateMatrix(vector<vector<int>> &mat)
    {
        int n = mat.size(), m = mat[0].size();
        vector<vector<int>> vis(n, vector<int>(m, 0));
        vector<vector<int>> dis(n, vector<int>(m, 0));
        queue<pair<pair<int, int>, int>> ele;
        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < m; j++)
            {
                if (mat[i][j] == 0)
                {
                    ele.push({{i, j}, 0});
                    vis[i][j] = 1;
                }
                else
                {
                    vis[i][j] = 0;
                }
            }
        }
        int surrow[] = {-1, 0, +1, 0};
        int surcol[] = {0, +1, 0, -1};
        while (!ele.empty())
        {
            int row = ele.front().first.first;
            int col = ele.front().first.second;
            int step = ele.front().second;

            ele.pop();
            dis[row][col] = step;

            for (int i = 0; i < 4; i++)
            {
                int nrow = row + surrow[i];
                int ncol = col + surcol[i];

                if (nrow >= 0 && ncol >= 0 && nrow < n && ncol < m && vis[nrow][ncol] == 0)
                {
                    vis[nrow][ncol] = 1;
                    ele.push({{nrow, ncol}, step + 1});
                }
            }
        }

        return dis;
    }
};
// @lc code=end

File: Java/Grind75/Week 3/542.01-matrix.java
import java.util.*;

/*
 * @lc app=leetcode id=542 lang=java
 *
 * [542] 01 Matrix
 */

// @lc code=start
class Solution {
    public int[][] updateMatrix(int[][] mat) {
        int height = mat.length, width = mat[0].length;
        int[][] dist = new int[height][width], vis = new int[height][width];
        Queue<int[]> queue = new ArrayDeque<>();
        // Scanning for 0s
        for (int i = 0; i < height; i++) {
            for (int j = 0; j < width; j++) {
                if (mat[i][j] == 0) {
                    vis[i][j] = 1;
                    dist[i][j] = 0;
                    queue.offer(new int[] { i, j, 0 });
                }
            }
        }
        // bfs
        while (!queue.isEmpty()) {
            int[] node = queue.poll();
            int[] dirs = { 1, 0, -1, 0, 1 };
            for (int i = 0; i < dirs.length - 1; i++) {
                int row = node[0] + dirs[i];
                int col = node[1] + dirs[i + 1];
                if (row >= 0 && row < height && col >= 0 && col < width && vis[row][col] != 1) {
                    vis[row][col] = 1;
                    dist[row][col] = node[2] + 1;
                    queue.offer(new int[] { row, col, node[2] + 1 });
                }
            }
        }
        return dist;
    }
}
// @lc code=end
